### PR TITLE
update smart contract to use custom error instead of revert reason

### PR DESCRIPTION
**Pull request by chiro-hiro**

### CODE REVIEW

File: contracts/contracts/orand/OrandProviderV1.sol [hunk 1]
```diff
@@ -35,13 +36,27 @@ contract OrandProviderV1 is OrandStorage, OrandManagement, OrandSignatureVerifie
...
+    // Unable to add epoch to storage
+    if (!_addEpoch(receiverAddress, newEpoch)) {
+      revert UnableToAddNewEpoch(receiverAddress, newEpoch);
+    }
+    // Unable to forward randomness to receiver contract
+    if (!IOrandConsumerV1(receiverAddress).consumeRandomness(newEpoch.y)) {
+      revert UnableToForwardRandomness(receiverAddress, y);
+    }
+    // Increasing nonce of receiver to prevent replay attack
+    if (!_increaseNonce(receiverAddress)) {
```
Comment (vuvoth) [inline contracts/contracts/orand/OrandProviderV1.sol:57]: I think something wrong in this logic. `_increaseNonce` always return true.

File: contracts/contracts/orand/OrandStorage.sol [hunk 2]
```diff
@@ -1,47 +1,28 @@
...
 
   // Total number of epoch
-  mapping(address => uint256) totalEpoch;
+  mapping(address => uint256) internal totalEpoch;
 
   // Check if epoch is valid
   modifier onlyValidEpoch(address receiverAddress, uint256 epoch) {
-    require(epoch < totalEpoch[receiverAddress] && epoch >= 1, 'OS: Invalid epoch id');
-    require(storageEpoch[receiverAddress][epoch].sued == 0, 'OS: This epoch was sued');
+    if (epoch >= totalEpoch[receiverAddress] && epoch < 1) {
```
Comment (vuvoth) [inline contracts/contracts/orand/OrandStorage.sol:20]:
```suggestion
    if (epoch >= totalEpoch[receiverAddress] || epoch < 1) {
```

File: contracts/tasks/deploy-orandv1.ts [hunk 3]
```diff
@@ -10,23 +10,19 @@ import { OrandProviderV1 } from '../typechain-types/OradProvider.sol';
...
-      ['Bytes', 'Verifier'],
+      [],
+      // This public key is corresponding to 0x7e9e03a453867a7046B0277f6cD72E1B59f67a0e
       [
         '0x46b01e9550b56f3655dbca90cfe6b31dec3ff137f825561c563444096803531e',
         '0x9d4f6e8329d300483a919b63843174f1fca692fc6d2c07b985f72386e4edc846',
       ],
       // Operator address
-      accounts[0].address,
+      '0x7e9e03a453867a7046B0277f6cD72E1B59f67a0e',
```
Comment (vuvoth) [inline contracts/tasks/deploy-orandv1.ts:25]: I think we should save this address account and public keys to env file.

File: contracts/contracts/orand/OrandProviderV1.sol (hunk 1 above)
Comment (chiro-hiro) [reply contracts/contracts/orand/OrandProviderV1.sol:57]: Just want to handle unexpected case, it looks unnecessary but it made me feel safe.

```solidity
require(_increaseNonce(receiverAddress), "It suppose to work");
```

File: contracts/contracts/orand/OrandStorage.sol (hunk 2 above)
Comment (chiro-hiro) [reply contracts/contracts/orand/OrandStorage.sol:20]: Nice catch, I was wrong when I try to convert to inverted logic

File: contracts/tasks/deploy-orandv1.ts (hunk 3 above)
Comment (chiro-hiro) [reply contracts/tasks/deploy-orandv1.ts:25]: Yeb, I will crease a following issue.

**Pull request merged**